crates/core/tests/integration.rs

```diff
@@ -0,0 +1,228 @@
+            airdrop_token_amount,
+            ..SimnetConfig::default()
+        },
+        rpc: RpcConfig {
+            bind_host: bind_host.to_string(),
+            bind_port,
+            ..Default::default()
+        },
+        ..SurfpoolConfig::default()
+    };
+
+    let (simnet_events_tx, simnet_events_rx) = unbounded();
+    let (_simnet_commands_tx, simnet_commands_rx) = unbounded();
+
+    let _handle = hiro_system_kit::thread_named("test").spawn(move || {
+        let future = start(config, simnet_events_tx, simnet_commands_rx);
+        if let Err(e) = hiro_system_kit::nestable_block_on(future) {
+            panic!("{e:?}");
+        }
+    });
+
+    loop {
+        match simnet_events_rx.recv() {
+            Ok(SimnetEvent::Ready) => break,
+            _ => (),
+        }
+    }
+
+    let minimal_client =
+        http::connect::<MinimalClient>(format!("http://{bind_host}:{bind_port}").as_str())
+            .await
+            .expect("Failed to connect to Surfpool");
+    let full_client =
+        http::connect::<FullClient>(format!("http://{bind_host}:{bind_port}").as_str())
+            .await
+            .expect("Failed to connect to Surfpool");
+
+    let recent_blockhash = full_client
+        .get_latest_blockhash(None)
+        .await
+        .map(|r| {
+            Hash::from_str(r.value.blockhash.as_str()).expect("Failed to deserialize blockhash")
+        })
+        .expect("Failed to get blockhash");
+
+    let balances = join_all(
+        airdrop_addresses
+            .iter()
+            .map(|pk| minimal_client.get_balance(pk.to_string(), None)),
+    )
+    .await
+    .into_iter()
+    .collect::<Result<Vec<_>, _>>()
+    .expect("Failed to fetch balances");
+
+    assert!(
+        balances.iter().all(|b| b.value == airdrop_token_amount),
+        "All addresses did not receive the airdrop"
+    );
+
+    let _transfers = join_all(airdrop_keypairs.iter().map(|kp| {
+        let msg = Message::new_with_blockhash(
+            &[system_instruction::transfer(
+                &kp.pubkey(),
+                &airdrop_addresses[0],
+                airdrop_token_amount / 2,
+            )],
+            Some(&kp.pubkey()),
+            &recent_blockhash,
+        );
+
+        let Ok(tx) = VersionedTransaction::try_new(
+            VersionedMessage::Legacy(msg),
+            &vec![kp.insecure_clone()],
+        ) else {
+            return Box::pin(future::err(Error::invalid_params("tx")));
+        };
+
+        let Ok(encoded) = bincode::serialize(&tx) else {
+            return Box::pin(future::err(Error::invalid_params("encoded")));
+        };
+        let data = bs58::encode(encoded).into_string();
+
+        Box::pin(future::ready(Ok(full_client.send_transaction(data, None))))
+    }))
+    .await
+    .into_iter()
+    .collect::<Result<Vec<_>, _>>()
+    .expect("Transfers failed");
+
+    // Wait for all transactions to be received
+    // let mut processed = 0;
```

Comment:
Is there a chance that because we've `await`ed above, one of them has already emitted the processed event, so we're too late when we start the loop below?

Maybe we need to start that loop in a thread _before_ sending the transfers, and the test will fail if that thread isn't closed within a certain amount of time?

Reply:
Sorry I forgot to ping: @Dodecahedr0x